### PR TITLE
[SPARK-31242][SQL][TEST] mergeSparkConf in WithTestConf should also respect spark.sql.legacy.sessionInitWithConfigDefaults

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

Make `mergeSparkConf` in `WithTestConf` respects `spark.sql.legacy.sessionInitWithConfigDefaults`.

### Why are the changes needed?

Without the fix, conf specified by `withSQLConf` can be reverted to original value in a cloned SparkSession.  For example, you will fail test below without the fix:

```
withSQLConf(SQLConf.CODEGEN_FALLBACK.key -> "true") {
  val cloned = spark.cloneSession()
  SparkSession.setActiveSession(cloned)
  assert(SQLConf.get.getConf(SQLConf.CODEGEN_FALLBACK) === true)
}
```

So we should fix it just as  #24540 did before.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added tests.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -780,5 +780,14 @@ class AdaptiveQueryExecSuite
       }
     }
   }
+
+  test("SPARK-31242: clone SparkSession should respect sessionInitWithConfigDefaults") {
```
Comment (Ngone51) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala]: I know it's not directly related to AQE. But the issue can not be reproduced without `SharedSparkSessionBase` and seems AQE now is more likely related to `cloneSession`. So I decide to add it here.

**Comment by Ngone51** [comment]
cc @cloud-fan @jose-torres

**Comment by Ngone51** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala]: how about `SQLQuerySuite`?

Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala]: sounds good.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by Ngone51** [comment]
thanks!

**Pull request closed without merging**